src/main/java/uk/gov/ons/census/action/schedule/ActionRuleProcessor.java

```diff
@@ -96,23 +96,31 @@ private void executeClassifiedCases(ActionRule triggeredActionRule) {
   }
 
   private void executeCases(Stream<Case> cases, ActionRule triggeredActionRule) {
-    List<Callable<Boolean>> callables = new LinkedList<>();
+    List<Callable<ActionInstruction>> callables = new LinkedList<>();
     cases.forEach(
         caze -> {
           callables.add(
-              () -> {
-                createAndSendActionRequest(caze, triggeredActionRule);
-                return Boolean.TRUE;
-              });
+              () -> actionInstructionBuilder.buildActionInstruction(caze, triggeredActionRule));
         });
 
     try {
-      List<Future<Boolean>> results = EXECUTOR_SERVICE.invokeAll(callables);
-
-      for (Future<Boolean> result : results) {
-        if (result.get() != Boolean.TRUE) {
-          throw new RuntimeException(); // One of the threads had a problem
+      final String routingKey =
+          String.format(
+              "%s%s%s",
+              ROUTING_KEY_PREFIX,
+              triggeredActionRule.getActionType().getHandler(),
+              ROUTING_KEY_SUFFIX);
+
+      List<Future<ActionInstruction>> results = EXECUTOR_SERVICE.invokeAll(callables);
+
+      log.info("About to send {} ActionInstruction messages", results.size());
+      int messagesSent = 0;
+      for (Future<ActionInstruction> result : results) {
+        if (messagesSent++ % 1000 == 0) {
+          log.info("Sent {} ActionInstruction messages", messagesSent);
         }
+
+        rabbitTemplate.convertAndSend(outboundExchange, routingKey, result.get());
       }
     } catch (InterruptedException e) {
```

Comment:
this could be combined with the exception below, 2 lines of code saved.   (InterruptedException | ExecutionException e)

Reply:
eye roll